### ISSUE TITLE
security overrides custom attributes

### ISSUE THREAD

**Issue by joukewitteveen** [asciidoctor/asciidoctor#3563]
In [`lib/asciidoctor/document.rb#L427-L431`](https://github.com/asciidoctor/asciidoctor/blob/v2.0.10/lib/asciidoctor/document.rb#L427-L431), security settings are enforced by modifying `docdir` and `docfile`. This is documented in https://asciidoctor.org/docs/user-manual/#running-asciidoctor-securely.

As a result, it is also impossible to set `docdir` and `docfile` manually when converting a string.
```ruby
::Asciidoctor.convert(input_string, attributes: { 'docdir': '.', 'docfile': 'README' })
```
Is this intended behavior? I would argue that attributes that are set by the calling party should be honored.

Additionally, there is a minor bug exposed by the case described above. In [document.rb](https://github.com/asciidoctor/asciidoctor/blob/v2.0.10/lib/asciidoctor/document.rb#L427-L431), we find
```ruby
        attr_overrides['docfile'] = attr_overrides['docfile'][(attr_overrides['docdir'].length + 1)..-1]
```
which transforms `docfile` into `ADME`.

**Comment by mojavelinux** [comment]
You need to set the safe mode when using the API.

**Comment by joukewitteveen** [comment]
Yeah, I know. The code example was meant to convey the idea, but indeed, it should have been something like
```ruby
::Asciidoctor.convert(input_string, safe: :secure, attributes: { 'docdir': '.', 'docfile': 'README' })
```

Anyhow, the issue still stands: the attributes are not preserved, even when they are set by the calling party and not just inferred by `asciidoctor`.

**Comment by joukewitteveen** [comment]
To be clear: this is not intended to be a support case. This is intended as
1. A _feature request_: Preserve externally set attributes (`docdir`, `docfile`) regardless of safe mode
2. A _bug report_: Mangling of `docfile` is incorrect when `docdir` and `docfile` are set externally

Implementing the feature request will automatically fix the reported bug.

**Comment by mojavelinux** [comment]
I'm not sure if it's intentional, but you have a mistake in the API call. You are passing the attribute as symbols instead of a strings.

It should be:

```rb
Asciidoctor.convert '{docfile}', safe: :secure, attributes: { 'docdir' => '.', 'docfile' => 'README.adoc' }
```

I do agree that docfile seems to be getting mangled and docdir is always suppressed. It's a reasonable request.

**Comment by mojavelinux** [comment]
I've updated the title to be more clear.

**Comment by joukewitteveen** [comment]
Thanks for acknowledging the issue and sorry for botching up my example not once, but twice!

In this context, it would be nice if the [documentation](https://asciidoctor.org/docs/user-manual/#running-asciidoctor-securely) contained a sentence or two about _why_ `docfile` and `docdir` are unsafe to expose. If I knew the reason, I would have opened a PR, but I don't.

**Comment by mojavelinux** [comment]
No problem. Thanks for following up.

> In this context, it would be nice if the documentation contained a sentence or two about why docfile and docdir are unsafe to expose.

On a server, it would be possible to reveal information about the filesystem on which the application is running using `{docfile}` and `{docdir}`. That's why they are masked. However, if the API sets these attributes explicitly, we'll assume that's the intended behavior.

If you don't want them masked, you can set the safe mode to `:safe` instead of `:secure`.

**Comment by mojavelinux** [comment]
I meant to state that I agree that if the docfile and docdir are locked (which means they are set from the API or CLI), then they should not be masked. We'll assume that these assignments are intentional.

**Comment by mojavelinux** [comment]
> To be clear: this is not intended to be a support case.

To provide some context here, I mark issues as "support" until I can determine that a code change is required. This isn't meant to downplay the report in any way, but rather help me manage the issues so I know which ones are ready for a code change.

**Comment by mojavelinux** [comment]
>  there is a minor bug exposed by the case described above, which transforms docfile into ADME.

The assumption is that docfile includes docdir (and that both are absolute).

**Comment by mojavelinux** [comment]
This turns out to be *a lot* more complicated than I anticipated. There are lot of different code paths to consider, and it's not immediately clear what information needs to be passed where to know when docfile and docdir have been overridden (file vs string input, relationship to base_dir, etc). We'll need to defer until at least 2.1 to study more carefully.

**Comment by mojavelinux** [comment]
I'm thinking we might need to pass in file and dir to the Document constructor using a separate option so we know where the information is coming from, then assign the docfile and docdir attributes accordingly. Just nothing that idea for my future self.
